### DISCUSSION TITLE
How to use jotai?

### DISCUSSION THREAD

**Discussion by pruge** [pmndrs/jotai#2311] [Q&A] [answered]
I'm a newbie.
I would like to take one of the object arrays, modify it, and reflect it.
However, it doesn't work as intended.

Here is an example:
[sandbox](https://codesandbox.io/p/sandbox/jotai-derived-atom-render-count-test-forked-jzpwnn?file=%2Fsrc%2FApp.jsx%3A31%2C1)

```js
const keyAtom = atom("a");
const countsAtom = atom([{ key: "a", count: 1 }]);
const derivedCountAtom = atom((get) => {
  const key = get(keyAtom);
  const counts = get(countsAtom);
  const count = find(counts, (count) => count.key == key);

  return count;
});
const updateCountAtom = atom(null, (get, set, update) => {
  const counts = get(countsAtom);
  const idx = findIndex(counts, (count) => count.key === update.key);
  counts[idx] = update;
  set(countsAtom, counts);
});
```

If it is used incorrectly, how should I correct it?

**Comment by rothsandro** [comment, marked as answer]
In your setter function, you mutate the existing state and then provide it to the `set()` function. Jotai will only re-render the component if the state has actually changed, which isn't the case because you provided the same object.

```js
counts[idx] = update,    // <-- Mutates the existing state
set(countsAtom, counts); // <-- Sets the existing, mutated object
```

When using arrays/objects, you should never mutate them but instead create a new object. You can use `slice()` to generate a shallow copy of the array. For the objects inside the array you already create a new object so that's okay.

```diff
const updateCountAtom = atom(null, (get, set, update) => {
-  const counts = get(countsAtom);
+  const counts = get(countsAtom).slice();
  // ...
});
```

For more complex cases you can also use something like [Immer](https://jotai.org/docs/extensions/immer).

**Reply by pruge** [reply]
I tried it using immer.
It works as intended. thank you

```js
const countsAtom = atomWithImmer([{ key: "a", count: 1 }]);
const updateCountAtom = atom(null, (get, set, update) => {
  set(countsAtom, (counts) => {
    const idx = findIndex(counts, (count) => count.key === update.key);
    counts[idx] = update;
  });
});
```

[sandbox](https://codesandbox.io/p/sandbox/jotai-derived-atom-render-count-test-forked-jzpwnn?file=%2Fsrc%2FApp.jsx%3A7%2C1-21%2C1)
